### PR TITLE
[Generator] Refactor Format Type methods into TypeManager class

### PR DESCRIPTION
**Pull request by dustin-wojciechowski**
First take at refactoring parts of the Generator:

1. RemoveArity() is now a string extension method with a test
2. FormatType, FormatTypeUsedIn, and part of PrimitiveType methods have been moved into the TypeManager class.
3. TypeManager now has access to BindingTouch (similar to the other Manager classes) so that it may call methods from NamespaceManager, which also had to be made into a public property of BindingTouch vs being passed through the Generator constructor.

### CODE REVIEW

**Review by mandel-macaque — APPROVED** [review]

File: src/bgen/TypeManager.cs [hunk 1]
```diff
@@ -413,4 +418,120 @@ public TypeManager (BindingTouch bindingTouch, Assembly apiAssembly, Assembly co
...
+
+		if (type.IsArray)
+			return FormatTypeUsedIn (usedInNamespace, type.GetElementType ()) + "[" + new string (',', type.GetArrayRank () - 1) + "]";
+
+		var interfaceTag = protocolized == true ? "I" : "";
+		string tname;
+		// we are adding the usage of ReflectedType just for those cases in which we have nested enums/classes, this soluction does not
+		// work with nested/nested/nested classes. But we are not writing a general solution because:
+		// 1. We have only encountered nested classes.
+		// 2. We are not going to complicate the code more than needed if we have never ever faced a situation with a crazy complicated nested hierarchy,
```
Comment (dalexsoto) [inline src/bgen/TypeManager.cs:514]: Fixes: https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1907601

```suggestion
		// 2. We are not going to complicate the code more than needed if we have never ever faced a situation with a super complicated nested hierarchy, 
```

**Review by dalexsoto — CHANGES_REQUESTED** [review]
Compliance :) https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1907601

**Review by haritha-mohan — APPROVED** [review]

File: src/bgen/TypeManager.cs [hunk 2]
```diff
@@ -1,11 +1,15 @@
...
 using System.Collections.Generic;
 using System.Diagnostics.CodeAnalysis;
+using System.Linq;
 using System.Reflection;
 
 #nullable enable
 
 public class TypeManager {
+	public BindingTouch BindingTouch;
+	readonly HashSet<string> typesThatMustAlwaysBeGloballyNamed = new ();
```
Comment (rolfbjarne) [inline src/bgen/TypeManager.cs:11]: Nothing seems to ever add to this hash set?

**Review by rolfbjarne — CHANGES_REQUESTED** [review]

File: src/bgen/TypeManager.cs (hunk 2 above)
Comment (dustin-wojciechowski) [reply src/bgen/TypeManager.cs]: Ah, I forgot to bring over that logic from the Generator. Added in new commit.

File: src/bgen/TypeManager.cs [hunk 3]
```diff
@@ -413,4 +433,124 @@ public TypeManager (BindingTouch bindingTouch, Assembly apiAssembly, Assembly co
...
+			tname = name;
+		else
+			tname = "global::" + @namespace + "." + name;
+
+		return tname;
+	}
+
+	public string FormatTypeUsedIn (string? usedInNamespace, Type? type, bool protocolized = false)
+	{
+		if (type is null)
```
Comment (dustin-wojciechowski) [inline src/bgen/TypeManager.cs:482]: I added a condition for null on account of type.GetElementType() (called in 524) having the possibility of null values (example in the mono docs even have a null example, interesting)

File: src/bgen/BindingTouch.cs [hunk 4]
```diff
@@ -69,9 +69,11 @@ public class BindingTouch : IDisposable {
 
 	public MetadataLoadContext? universe;
 	public TypeManager? typeManager;
+	NamespaceManager? namespaceManager;
 	public TypeManager TypeManager => typeManager!;
 	public Frameworks? Frameworks;
 	public AttributeManager? AttributeManager;
+	public NamespaceManager NamespaceManager => namespaceManager!;
```
Comment (haritha-mohan) [inline src/bgen/BindingTouch.cs:76]: nit but having the similar types grouped together makes it easier to follow

```suggestion
	public TypeManager? typeManager;
	public TypeManager TypeManager => typeManager!;
	public Frameworks? Frameworks;
	public AttributeManager? AttributeManager;
	public NamespaceManager NamespaceManager => namespaceManager!;
	NamespaceManager? namespaceManager;
```

**Review by mandel-macaque — APPROVED** [review]

File: src/bgen/BindingTouch.cs (hunk 4 above)
Comment (dustin-wojciechowski) [reply src/bgen/BindingTouch.cs:76]: Made some of those changes in the next PR: https://github.com/xamarin/xamarin-macios/pull/19380

**Pull request merged**